File: workflow-sdk-node/src/runner/kafka_runner.ts
import { KafkaJS } from '@confluentinc/kafka-javascript';
const { Kafka } = KafkaJS;

import { EngineClient, Job } from '../client/types';
import { HandlerRegistry } from '../handler/registry';
import { DeduplicationCache } from './deduplication_cache.js';

// @ts-ignore: JobDispatchEvent is generated during build
import { JobDispatchEvent } from '../generated/workflow/v1/engine';

export interface KafkaRunnerConfig {
  workerId: string;
  brokers: string[];
  clientId?: string;
  dedupWindowMs?: number;
}

export class KafkaRunner {
  private kafka: KafkaJS.Kafka;
  private consumers: KafkaJS.Consumer[] = [];
  private dedupCache: DeduplicationCache;
  private running = false;

  constructor(
    private config: KafkaRunnerConfig,
    private engine: EngineClient,
    private registry: HandlerRegistry
  ) {
    this.kafka = new Kafka({
      kafkaJS: {
        clientId: config.clientId || config.workerId,
        brokers: config.brokers,
      }
    });
    this.dedupCache = new DeduplicationCache(config.dedupWindowMs);
  }

  async start(): Promise<void> {
    this.running = true;
    const jobTypes = this.registry.jobTypes();
    console.log(`KafkaRunner starting: workerId=${this.config.workerId} jobTypes=${jobTypes}`);

    for (const jobType of jobTypes) {
      await this.startConsumer(jobType);
    }
  }

  private async startConsumer(jobType: string): Promise<void> {
    const topic = `workflow.jobs.${jobType}`;
    const groupId = `workflow.workers.${jobType}`;

    const consumer = this.kafka.consumer({
      kafkaJS: {
        groupId,
        partitionAssigners: [KafkaJS.PartitionAssigners.cooperativeSticky],
        fromBeginning: true,
      }
    });
    this.consumers.push(consumer);

    await consumer.connect();
    await consumer.subscribe({ topics: [topic] });

    await consumer.run({
      eachBatch: async (payload: KafkaJS.EachBatchPayload) => {
        for (const record of payload.batch.messages) {
          if (!this.running) break;
          if (record.value) {
            try {
              await this.dispatch(record.value);
            } catch (err) {
              console.error(`Error dispatching record from ${topic}:`, err);
            }
          }
        }
        await payload.resolveOffset(payload.batch.messages[payload.batch.messages.length - 1].offset);
        await payload.commitOffsetsIfNecessary();
      },
    });
  }

  private async dispatch(data: Buffer): Promise<void> {
    const event = JobDispatchEvent.decode(data);

    if (event.dedupId && this.dedupCache.seenRecently(event.dedupId)) {
      console.debug(`kafkarunner: skipping duplicate job_id=${event.jobId} dedup_id=${event.dedupId}`);
      return;
    }

    console.info(`kafkarunner: dispatching job_id=${event.jobId} job_type=${event.jobType}`);

    const handler = this.registry.get(event.jobType);
    if (!handler) {
      console.warn(`kafkarunner: no handler for jobType=${event.jobType} jobId=${event.jobId}`);
      await this.engine.failJob(event.jobId, this.config.workerId, `no handler registered for ${event.jobType}`, false);
      return;
    }

    let variables: Record<string, unknown> = {};
    if (event.jobPayload && event.jobPayload.length > 0) {
      try {
        variables = JSON.parse(Buffer.from(event.jobPayload).toString());
      } catch (e) {
        console.error(`kafkarunner: failed to parse job payload for ${event.jobId}:`, e);
      }
    }

    const job: Job = {
      id: event.jobId,
      jobType: event.jobType,
      instanceId: event.instanceId,
      stepExecutionId: event.stepExecutionId,
      retriesRemaining: event.retriesRemaining,
      variables,
    };

    try {
      const ctx = HandlerRegistry.contextFrom(job);
      const result = await handler(ctx);
      console.info(`kafkarunner: job completed job_id=${job.id}`);
      await this.engine.completeJob(job.id, this.config.workerId, result.variablesToSet || {});
    } catch (err: any) {
      const retryable = !(err.name === 'NonRetryableError' || err.isNonRetryable);
      if (retryable) {
        console.warn(`kafkarunner: job failed (retryable) job_id=${job.id}:`, err);
      } else {
        console.warn(`kafkarunner: job failed (non-retryable) job_id=${job.id}:`, err);
      }
      await this.engine.failJob(job.id, this.config.workerId, err.message, retryable);
    }
  }

  async stop(): Promise<void> {
    this.running = false;
    for (const consumer of this.consumers) {
      await consumer.disconnect();
    }
    this.dedupCache.close();
  }
}

File: workflow-sdk-python/src/workflow_sdk/runner/kafka_runner.py
import json
import logging
import threading
import signal
from typing import List, Optional, Dict, Any
from confluent_kafka import Consumer, KafkaError, KafkaException

from workflow_sdk.client.interface import EngineClient
from workflow_sdk.handler.registry import HandlerRegistry
from workflow_sdk.runner.deduplication_cache import DeduplicationCache
from workflow_sdk.errors import NonRetryableError

# Try to import generated proto, fallback to a dummy if not found during implementation
try:
    from workflow_sdk.internal.gen import engine_pb2
except ImportError:
    # Dummy class to avoid immediate crash if stubs aren't generated yet
    class engine_pb2:
        class JobDispatchEvent:
            @staticmethod
            def IsDefined(): return False

log = logging.getLogger(__name__)

class KafkaRunner:
    def __init__(
        self,
        worker_id: str,
        brokers: str,
        client: EngineClient,
        registry: HandlerRegistry,
        dedup_window_seconds: float = 600.0,
        extra_kafka_config: Optional[Dict[str, Any]] = None
    ):
        self.worker_id = worker_id
        self.brokers = brokers
        self.client = client
        self.registry = registry
        self.dedup_cache = DeduplicationCache(dedup_window_seconds)
        self.running = False
        self.consumers: List[Consumer] = []
        
        self.kafka_config = {
            'bootstrap.servers': brokers,
            'auto.offset.reset': 'earliest',
            'enable.auto.commit': False,
            'partition.assignment.strategy': 'cooperative-sticky',
        }
        if extra_kafka_config:
            self.kafka_config.update(extra_kafka_config)

    def run(self):
        self.running = True
        job_types = self.registry.job_types()
        log.info("KafkaRunner starting: workerId=%s jobTypes=%s", self.worker_id, job_types)

        threads = []
        for job_type in job_types:
            t = threading.Thread(target=self._consume_loop, args=(job_type,), name=f"kafka-consumer-{job_type}")
            t.start()
            threads.append(t)

        # Handle termination signals
        def handle_sig(sig, frame):
            log.info("Signal received, stopping...")
            self.stop()

        signal.signal(signal.SIGINT, handle_sig)
        signal.signal(signal.SIGTERM, handle_sig)

        for t in threads:
            t.join()
        
        self.dedup_cache.close()

    def stop(self):
        self.running = False

    def _consume_loop(self, job_type: str):
        topic = f"workflow.jobs.{job_type}"
        group = f"workflow.workers.{job_type}"
        
        conf = self.kafka_config.copy()
        conf['group.id'] = group
        
        consumer = Consumer(conf)
        self.consumers.append(consumer)
        
        try:
            consumer.subscribe([topic])
            
            while self.running:
                msg = consumer.poll(timeout=0.5)
                if msg is None:
                    continue
                
                if msg.error():
                    if msg.error().code() == KafkaError._PARTITION_EOF:
                        continue
                    else:
                        log.error("Kafka error: %s", msg.error())
                        continue

                try:
                    self._dispatch(msg.value())
                    consumer.commit(msg)
                except Exception as e:
                    log.exception("Error dispatching Kafka record from %s: %s", topic, e)
        finally:
            consumer.close()

    def _dispatch(self, data: bytes):
        if not hasattr(engine_pb2, 'JobDispatchEvent'):
            log.error("JobDispatchEvent proto not found. Cannot decode Kafka message.")
            return

        event = engine_pb2.JobDispatchEvent()
        event.ParseFromString(data)
        
        if event.dedup_id and self.dedup_cache.seen_recently(event.dedup_id):
            log.debug("kafkarunner: skipping duplicate job_id=%s dedup_id=%s", event.job_id, event.dedup_id)
            return

        log.info("kafkarunner: dispatching job_id=%s job_type=%s", event.job_id, event.job_type)

        handler = self.registry.get(event.job_type)
        if not handler:
            log.warning("kafkarunner: no handler for jobType=%s jobId=%s", event.job_type, event.job_id)
            self.client.fail_job(event.job_id, self.worker_id, f"no handler registered for {event.job_type}", False)
            return

        variables = {}
        if event.job_payload:
            try:
                variables = json.loads(event.job_payload.decode('utf-8'))
            except Exception as e:
                log.error("kafkarunner: failed to parse job payload for %s: %s", event.job_id, e)

        ctx = {
            "id": event.job_id,
            "jobType": event.job_type,
            "instanceId": event.instance_id,
            "stepExecutionId": event.step_execution_id,
            "retriesRemaining": event.retries_remaining,
            "variables": variables,
        }

        try:
            result = handler(ctx)
            log.info("kafkarunner: job completed job_id=%s", event.job_id)
            self.client.complete_job(event.job_id, self.worker_id, result or {})
        except NonRetryableError as e:
            log.warning("kafkarunner: job failed (non-retryable) job_id=%s: %s", event.job_id, e)
            self.client.fail_job(event.job_id, self.worker_id, str(e), False)
        except Exception as e:
            log.exception("kafkarunner: job failed (retryable) job_id=%s: %s", event.job_id, e)
            self.client.fail_job(event.job_id, self.worker_id, str(e), True)
